archive/imo/imo1960_q1.lean

```diff
@@ -0,0 +1,220 @@
+/-
+Copyright (c) 2020 Kevin Lacker. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Lacker
+-/
+
+import data.nat.digits
+
+open nat
+
+/-!
+# IMO 1960 Q1
+
+Determine all three-digit numbers $N$ having the property that $N$ is divisible by 11, and
+$\dfrac{N}{11}$ is equal to the sum of the squares of the digits of $N$.
+
+Since Lean doesn't have a way to directly express problem statements of the form
+"Determine all X satisfying Y", we express two predicates where proving that one implies the
+other is equivalent to solving the problem. A human solver also has to discover the
+second predicate.
+
+In theory, we should be able to brute force this and check every number up to 1000.
+In practice, the Lean tactics that are capable of checking a single number, like `norm_num`,
+are not efficient enough to run a thousand times while compiling one file.
+-/
+
+def sum_of_squares (L : list ℕ) : ℕ := L.foldr (λ x y, x*x + y) 0
+
+def problem_predicate (n : ℕ) :=
+(nat.digits 10 n).length = 3 ∧ 11 ∣ n ∧ n / 11 = sum_of_squares (nat.digits 10 n)
+
+def solution_predicate (n : ℕ) := n = 550 ∨ n = 803
+
+/-
+Proving that three digit numbers are the ones in [100, 1000).
+-/
+
+lemma not_zero (n : ℕ) (h1 : problem_predicate n) : n ≠ 0 :=
+have h2: nat.digits 10 n ≠ list.nil, from list.ne_nil_of_length_eq_succ h1.left,
+digits_ne_nil_iff_ne_zero.mp h2
+
+lemma ge_100 (n : ℕ) (h1 : problem_predicate n) : n ≥ 100 :=
+have h2 : 10^3 ≤ 10 * n, from begin
+  rw ← h1.left,
+  refine nat.base_pow_length_digits_le 10 n _ _,
+  simp,
+  exact not_zero n h1,
```

Comment:
(the suggestion blocks got a bit garbled on github, hopefully it makes sense in context) As there are 2 goals mathlib style is to bracket each goal separately like
```suggestion
  { simp, },
  { exact not_zero n h1, },
```
to aid a readers understanding of the proof.
In this case however the `exact ...` can be moved into the refine
```suggestion
  refine nat.base_pow_length_digits_le 10 n _ (not_zero n h1),
  simp,
```

Reply:
ok